Cable Damage Update

OmegaHaxors opened the pull request:
Revamped the way cables damage entities.
Cables no longer hurt non-living creatures.
Touching a cable will hurt a lot and send you flying backwards.
Creepers shocked by cables will explode

OmegaHaxors committed:
Cable Damage Update

Revamped the way cables damage entities.
Cables no longer hurt non-living creatures.
Touching a cable will hurt a lot and send you flying backwards.
Creepers shocked by cables will explode

jrddunbr commented on src/main/java/org/ja13/eau/EAU.java:
```diff
@@ -135,7 +135,7 @@
     public final static String UPDATE_URL = "https://github.com/jrddunbr/EAU/releases";
     public final static String SRC_URL = "https://github.com/jrddunbr/EAU";
-    public final static String[] AUTHORS = {"jrddunbr"};
+    public final static String[] AUTHORS = {"jrddunbr,Omega Haxors"};
```
You and I are not the same person. This is a list of strings. Get your own darn string! :D

```java
public final static String[] AUTHORS = {"jrddunbr", "Omega Haxors"};
```

OmegaHaxors replied:
Bunk Beds!

jrddunbr commented on src/main/java/org/ja13/eau/misc/Coordonate.kt:
```diff
@@ -143,7 +144,22 @@ class Coordonate : INBTTReady {
     fun distanceTo(e: Entity): Double {
-        return Math.abs(e.posX - (x + 0.5)) + Math.abs(e.posY - (y + 0.5)) + Math.abs(e.posZ - (z + 0.5))
+        //return Math.abs(e.posX - (x + 0.5)) + Math.abs(e.posY - (y + 0.5)) + Math.abs(e.posZ - (z + 0.5))
+        return Math.sqrt(Math.pow((e.posX - (x+0.5)),2.0) + Math.pow((e.posY - (y+0.5)),2.0) + Math.pow((e.posZ - (z+0.5)),2.0))
```
This is really expensive compared to the previous function. I think we were going for speed over accuracy.

OmegaHaxors replied:
I tested the two methods and the previous one didn't cause any regressions so I went ahead returned it back to the cheaper method.

jrddunbr commented on src/main/java/org/ja13/eau/sixnode/electriccable/ElectricCable.kt:
```diff
@@ -192,20 +202,39 @@ class ElectricCableRender(tileEntity: org.ja13.eau.node.six.SixNodeEntity, side:
     override fun process(time: Double) {
         val harmLevel = Math.max(0.0, (electricalLoad.u - 50 - insulationVoltage) / 500.0)
         val objects = location.world().getEntitiesWithinAABB(Entity::class.java, location.getAxisAlignedBB(4))
+
         for(obj in objects) {
+
```
I see we are in the business of introducing random white space.

OmegaHaxors replied:
You don't want to know what used to be in that white space.

jrddunbr commented on src/main/java/org/ja13/eau/sixnode/electriccable/ElectricCable.kt:
```diff
@@ -192,20 +202,39 @@ class ElectricCableRender(tileEntity: org.ja13.eau.node.six.SixNodeEntity, side:
             val distance = location.distanceTo(ent)
             val pain = (harmFunction(distance) * harmLevel).toFloat()
-            if (distance < 3 && pain > 0.05) {
+            if(distance < 1 && pain > 0.05 && ent.onGround){ //if they're touching the block, deal extra effects
+                if(ent is EntityCreeper){ location.world().createExplosion(ent,ent.posX,ent.posY+0.3,ent.posZ,1f,true); ent.setDead() }
+                ent.hurtResistantTime = 0
                 ent.attackEntityFrom(DamageSource("Cable"), pain)
+                if(ent.isEntityAlive) {
+                    ent.setVelocity(Math.min(location.directionOf(ent).x*pain,1.0), Math.min(0.5*pain,3.0), Math.min((location.directionOf(ent).z*pain),1.0))
+                }
             }
+            else if (distance < 3 && pain > 0.05 && ent.hurtResistantTime == 0 && ent.onGround) { //otherwise deal less punishing effects
+                ent.attackEntityFrom(DamageSource("Cable"), pain)
+                ent.hurtResistantTime = 15
+                if(ent is EntityCreeper && pain > 1){ location.world().createExplosion(ent,ent.posX,ent.posY+0.3,ent.posZ,1f,true); ent.setDead() }
+                if(ent.isEntityAlive){
+                    ent.setVelocity(0.0,Math.min(pain/2.0,0.65),0.0) //makes the zapped entity jump straight up based on pain, within limit
+                }
+            }
+        }
```
TODO: Playtest.

jrddunbr reviewed (changes requested):
Some minor stuff. That distance to method in utils needs reviewing as to where it is used and how. It looks like it was a basic function which needs to be reconsidered for efficiency. Doing square roots is expensive though, so avoid that if possible.

OmegaHaxors committed:
Cable Damage Update fixes

ready for pulling

OmegaHaxors committed:
Improved Creeper Detonation

Now with custom kill message

OmegaHaxors committed:
New Armors

Adds a ton of new armors, the alloys Bronze and Electrum, and Silver Ingots

OmegaHaxors committed:
Revert "New Armors"

This reverts commit df1653a036c614ec5630ee492694f88933f91484.

OmegaHaxors committed:
Merge pull request #1 from jrddunbr/main

Update

OmegaHaxors committed:
Copper Armor Fix

Copper Armor no longer errendiously uses "iron" as its build material. Now it will use the stats of Leather Armor (because it is a very easy set to make) and, get this, can be repaired with Copper ingots.

jrddunbr commented on src/main/java/org/ja13/eau/misc/Coordonate.kt:
```diff
@@ -10,6 +9,8 @@ import net.minecraft.util.AxisAlignedBB
 import net.minecraft.world.World
 import net.minecraftforge.common.DimensionManager
+import org.lwjgl.util.vector.Vector3f
+import javax.vecmath.Vector3d
```
Is this really what you meant to import?

OmegaHaxors replied:
Vector3d is used in some math, but Vector3f isn't needed.

jrddunbr commented on src/main/java/org/ja13/eau/sixnode/electriccable/ElectricCable.kt:
```diff
@@ -192,6 +198,8 @@ class ElectricCableRender(tileEntity: org.ja13.eau.node.six.SixNodeEntity, side:
     }
 }
 
+
+
```
Still in the business of whitespace, I see.

jrddunbr commented on src/main/java/org/ja13/eau/registry/ItemRegistry.kt:
```diff
@@ -128,27 +128,29 @@ class ItemRegistry {
+            val copperMaterial = EnumHelper.addArmorMaterial("Copper", 10, intArrayOf(1, 3, 2, 1), 15)
+            copperMaterial.customCraftingMaterial = EAU.copperIngot.newItemStack().item
             run {
                 name = I18N.TR_NAME(I18N.Type.ITEM, "Copper Helmet")
-                EAU.helmetCopper = genericArmorItem(ArmorMaterial.IRON, 2, ArmourType.Helmet, "eau:textures/armor/copper_layer_1.png", "eau:textures/armor/copper_layer_2.png").setUnlocalizedName(name).setTextureName("eau:copper_helmet").setCreativeTab(EAU.itemTab) as ItemArmor
+                EAU.helmetCopper = genericArmorItem(copperMaterial, 2, ArmourType.Helmet, "eau:textures/armor/copper_layer_1.png", "eau:textures/armor/copper_layer_2.png").setUnlocalizedName(name).setTextureName("eau:copper_helmet").setCreativeTab(EAU.itemTab) as ItemArmor
                 GameRegistry.registerItem(EAU.helmetCopper, "Eln.$name")
                 GameRegistry.registerCustomItemStack(name, ItemStack(EAU.helmetCopper))
             }
             run {
                 name = I18N.TR_NAME(I18N.Type.ITEM, "Copper Chestplate")
-                EAU.plateCopper = genericArmorItem(ArmorMaterial.IRON, 2, ArmourType.Chestplate, "eau:textures/armor/copper_layer_1.png", "eau:textures/armor/copper_layer_2.png").setUnlocalizedName(name).setTextureName("eau:copper_chestplate").setCreativeTab(EAU.itemTab) as ItemArmor
+                EAU.plateCopper = genericArmorItem(copperMaterial, 2, ArmourType.Chestplate, "eau:textures/armor/copper_layer_1.png", "eau:textures/armor/copper_layer_2.png").setUnlocalizedName(name).setTextureName("eau:copper_chestplate").setCreativeTab(EAU.itemTab) as ItemArmor
                 GameRegistry.registerItem(EAU.plateCopper, "Eln.$name")
                 GameRegistry.registerCustomItemStack(name, ItemStack(EAU.plateCopper))
             }
             run {
                 name = I18N.TR_NAME(I18N.Type.ITEM, "Copper Leggings")
-                EAU.legsCopper = genericArmorItem(ArmorMaterial.IRON, 2, ArmourType.Leggings, "eau:textures/armor/copper_layer_1.png", "eau:textures/armor/copper_layer_2.png").setUnlocalizedName(name).setTextureName("eau:copper_leggings").setCreativeTab(EAU.itemTab) as ItemArmor
+                EAU.legsCopper = genericArmorItem(copperMaterial, 2, ArmourType.Leggings, "eau:textures/armor/copper_layer_1.png", "eau:textures/armor/copper_layer_2.png").setUnlocalizedName(name).setTextureName("eau:copper_leggings").setCreativeTab(EAU.itemTab) as ItemArmor
                 GameRegistry.registerItem(EAU.legsCopper, "Eln.$name")
                 GameRegistry.registerCustomItemStack(name, ItemStack(EAU.legsCopper))
             }
             run {
                 name = I18N.TR_NAME(I18N.Type.ITEM, "Copper Boots")
-                EAU.bootsCopper = genericArmorItem(ArmorMaterial.IRON, 2, ArmourType.Boots, "eau:textures/armor/copper_layer_1.png", "eau:textures/armor/copper_layer_2.png").setUnlocalizedName(name).setTextureName("eau:copper_boots").setCreativeTab(EAU.itemTab) as ItemArmor
+                EAU.bootsCopper = genericArmorItem(copperMaterial, 2, ArmourType.Boots, "eau:textures/armor/copper_layer_1.png", "eau:textures/armor/copper_layer_2.png").setUnlocalizedName(name).setTextureName("eau:copper_boots").setCreativeTab(EAU.itemTab) as ItemArmor
                 GameRegistry.registerItem(EAU.bootsCopper, "Eln.$name")
                 GameRegistry.registerCustomItemStack(name, ItemStack(EAU.bootsCopper))
```
Honestly, I don't think copper armor is very practical. I can't find any evidence that it would be a good armor material. I see a lot of iron and *steel* when I look online for armor.

OmegaHaxors replied:
This was a nerf to lower the durability and stats of copper down to Leather.
It used to be IRON, which was overpowered and made it require Iron ingots to repair. So yes, copper armor is crap. This update makes copper armor crap.

jrddunbr commented on src/main/java/org/ja13/eau/sixnode/electriccable/ElectricCable.kt:
```diff
@@ -201,11 +209,36 @@ class PlayerHarmer(val electricalLoad: org.ja13.eau.sim.ElectricalLoad, private
             val distance = location.distanceTo(ent)
             val pain = (harmFunction(distance) * harmLevel).toFloat()
-            if (distance < 3 && pain > 0.05) {
+            if(distance < 1 && pain > 0.05 && ent.onGround){ //if they're touching the block, deal extra effects
+                if(ent is EntityCreeper){
+                    location.world().createExplosion(ent, ent.posX, ent.posY, ent.posZ, 3f, true)
+                    ent.customNameTag = "Electrically-detonated Creeper"
+                    ent.setDead()
+                }
+                if(ent is EntityEnderman){ ent.setFire(1) } //enderman will teleport
+                ent.hurtResistantTime = 0
                 ent.attackEntityFrom(DamageSource("Cable"), pain)
+                if(ent.isEntityAlive) {
+                    ent.setVelocity(Math.min(location.directionOf(ent).x*pain,1.0), Math.min(0.5*pain,3.0), Math.min((location.directionOf(ent).z*pain),1.0))
+                }
             }
+            else if (distance < 3 && pain > 0.05 && ent.hurtResistantTime == 0 && ent.onGround) { //otherwise deal less punishing effects
+                ent.attackEntityFrom(DamageSource("Cable"), pain)
+                ent.hurtResistantTime = 15
+                if(ent is EntityCreeper){
+                    location.world().createExplosion(ent, ent.posX, ent.posY, ent.posZ, 3f, true)
+                    ent.customNameTag = "Electrically-detonated Creeper"
+                    ent.setDead()
+                }
+                if(ent is EntityEnderman){ ent.setFire(1) } //enderman will teleport
+                if(ent.isEntityAlive){
+                    ent.setVelocity(0.0,Math.min(pain/2.0,0.65),0.0) //Electricity makes you jump
+                }
+            }
+        }
```
The two sides of the if are a lot of duplicated code, I'd rather see it as one entry for code and to change the values where applicable based on the values. It will make this code execute faster, which is kinda important since it happens per tick for a lot of stuff.

OmegaHaxors replied:
I'll go ahead and optimize that.

OmegaHaxors replied:
Thank you very much for bringing this up. The new codebase is MUCH cleaner and i'm glad I put the energy into fixing it.

jrddunbr reviewed (changes requested):
Still some stuff that isn't quite there. I'm concerned about the performance of the player harmer functionality causing TPS problems.

OmegaHaxors committed:
Cable Damage Fixes

Optimized code
